src/NuGet.Core/NuGet.ProjectModel/PackageSpecWriter.cs

```diff
@@ -617,15 +618,18 @@ private static void SetCentralDependencies(IObjectWriter writer, ICollection<Cen
             writer.WriteObjectEnd();
         }
 
-        private static int GetHash(ICollection<CentralPackageVersion> items)
+        private static string GetHash(ICollection<CentralPackageVersion> items)
         {
-            var hashCode = new HashCodeCombiner();
-            foreach (var item in items)
+            using (var hashFunc = new Sha512HashFunction())
```

Comment:
I think the idea behind that might've been a perf optimization. 

I can only imagine the perf gains are now gone right? We can just add it to the original hashing function?

Reply:
Basically this whole `compressed` concept, might be something that needs to go away.

Reply:
> I can only imagine the perf gains are now gone right? We can just add it to the original hashing function?
> Basically this whole compressed concept, might be something that needs to go away.

Yeah, I didn't think that it matters but apparently it has a non-negligible effect. Surprisingly, it is not the hash function itself that is slow. In profiler I can see that the fact that centralised versions are being sorted and also the conversion of VersionRange to string makes the "non-compressed" path to be actually slow:
```
foreach (var dependency in centralPackageVersions.OrderBy(dep => dep.Name))
{
    writer.WriteNameValue(name: dependency.Name, value: dependency.VersionRange.ToNormalizedString());
}
```

I need to try more things to come up with final conclusion on how to handle this issue without sacrificing performance.

Reply:
I think we need to keep the "compressed" flag for now. Without it (https://github.com/NuGet/NuGet.Client/pull/4523/commits/7e53a5054eaa532e838bc63104749eb8af914ce1)  the overall performance of (when measuring time of `NuGet.ProjectModel.DependencyGraphSpec.GetHash()` is almost 100% worse.
My latest commit (https://github.com/NuGet/NuGet.Client/pull/4523/commits/48c37d818c061afa4d4fd579ea56ca17d7c4152e) makes it only 10% worse. I think that it is acceptable cost for fixing no-op scenario.

We could try making it much faster by avoiding calculating a hash value multiple times for the same project. For large solutions that would mean massive savings. But it is a nontrivial change to do so I would like to keep it out of scope of this PR.

Reply:
Let's keep it then. 

> We could try making it much faster by avoiding calculating a hash value multiple times for the same project

The trick here is that you can't guarantee the package spec hasn't been changed. 
Out of scope for this PR though. Thanks for validating that.